modules/actions/cow-amm/upsert-pools.ts

```diff
@@ -0,0 +1,149 @@
+import { Chain } from '@prisma/client';
+import { prisma } from '../../../prisma/prisma-client';
+import { fetchCowAmmData } from '../../sources/contracts';
+import { poolUpsertsUsd } from '../../sources/enrichers/pool-upserts-usd';
+import type { CowAmmSubgraphClient } from '../../sources/subgraphs';
+import type { ViemClient } from '../../sources/types';
+import { subgraphPoolUpsert } from '../../sources/transformers/subgraph-pool-upsert';
+import { formatUnits } from 'viem';
+
+/**
+ * Gets a list of pool ids and fetches the data from the subgraph and rpc, and then upserts into the database.
+ *
+ * @param ids List of pool ids to fetch and upsert
+ * @param viemClient
+ * @param cowAmmSubgraphClient
+ * @param chain
+ */
+export const upsertPools = async (
+    ids: string[],
+    viemClient: ViemClient,
+    cowAmmSubgraphClient: CowAmmSubgraphClient,
+    chain: Chain,
+) => {
+    const { pools } = await cowAmmSubgraphClient.Pools({ where: { id_in: ids } });
+
+    // Get onchain data for the pools
+    const blockNumber = await viemClient.getBlockNumber();
+    const onchainData = await fetchCowAmmData(
+        pools.map((pool) => pool.id),
+        viemClient,
+        blockNumber,
+    );
+
+    // Needed to get the token decimals for the USD calculations,
+    // Keeping it external, because we fetch these tokens in the upsert pools function
+    const allTokens = await prisma.prismaToken.findMany({
+        where: {
+            chain: chain,
+        },
+    });
+
+    // Add tokens if any are missing in the DB
+    const poolTokens = pools.map((pool) => pool.tokens).flat();
+    const missingTokens = poolTokens.filter(
+        (poolToken) => !allTokens.find((token) => token.address.toLowerCase() === poolToken.address.toLowerCase()),
+    );
+    if (missingTokens.length > 0) {
+        await prisma.prismaToken.createMany({
+            data: missingTokens.map((token) => ({
+                address: token.address,
+                decimals: token.decimals,
+                symbol: token.symbol,
+                name: token.name,
+                chain,
+            })),
+        });
+    }
+
+    // Get the data for the tables about pools
+    const dbPools = pools.map((poolData) =>
+        subgraphPoolUpsert(
```

Comment:
I think I would refactor this to a more generalized subgraphPoolUpsert that takes v3 subgraph pools, v2 subgraph pools and cow subgraph pools. But can also be left for the future, should add a TODO